### ISSUE TITLE
URI/URN representations for GitBOM

### ISSUE THREAD

**Issue by edwarnicke** [omnibor/spec#2]
Both the [SPDX](https://spdx.dev/) and [CycloneDX](https://cyclonedx.org/) SBOM communities have reached out to inquire about referencing GitBOM in their SBOMs.

- [CycloneDX expressed a desire for a URN scheme](https://openssf.slack.com/archives/C02D1JYQLPQ/p1644454418177419?thread_ts=1644395359.750189&cid=C02D1JYQLPQ)
- [SPDX expressed a desire for a URI scheme](https://youtu.be/wbhyW7nEhHg?t=1500)

[RFC 3986](https://www.rfc-editor.org/rfc/rfc3986.html) defines URIs.  URIs are then further classified as URLs (Uniform Resource Locators) and URNs (Uniform Resource Name).

[URN Syntax](https://www.rfc-editor.org/rfc/rfc2141) is of the form:

```
<URN> ::= "urn:" <NID> ":" <NSS>
```

where <NID> is the Namespace Identifier, and <NSS> is the Namespace  Specific String.

Its is likely true that the NSS should be the gitref/git object id of the urn as a hex encodes string.  It remains to be determined what the NID should be.

At first glance one might consider using [`swh:1:cnt` NSS from the Software Heritage Project](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html#uri-scheme).  This has the advantage that the [swh scheme is already registered with IANA](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml) and that the NSS part for `swh:1:cnt` has as its NSS the `sha1_git` (ie, gitref or git object id) of the contents.

There has already been discussion in the GitBOM community about migration to the sha256 version of GitBOM at some point in the future.  It is probably good to bake that flexibility into any NID choice.

One might also have considered using `git` as the NID, but [`git` is already registered with IANA](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml) for other purposes.

One possibility might be to define the NID as a descriptive: `gitoid:${hash}` such as `gitoid:sha1` or `gitoid:sha256` (or some variation thereof).

**Comment by edwarnicke** [comment]
@coderpatros I like the idea of using `gitbom` too.

One the sha256 discussion... would you be willing to open an issue on this spec repo to discuss?  We need a clear plan around it for sure.

Even if we *start* with sha256, we probably want to be forward flexible over time (for when sha256 get's broken).

Maybe something like `gitbom:${hash}:` like `gitbom:sha1` or `gitbom:sha256` ?

**Comment by edwarnicke** [comment]
@coderpatros Thinking a bit more about this... it might be handy to actually use a scheme that describes what we are talking about independently of gitbom, so that if others wish to use it they can.  What we are talking about is the gitoid for a type of git object using a particular hash.  So perhaps:

`gitoid:blob:sha1:${hash}` 

Thoughts

**Comment by kestewart** [comment]
@edwarnicke , @coderpatros - like the addition of including the hash in the formation, and yes, keeping it flexible to adapt crypto algs, is a good bit of future proofing. 

SPDX should be able to align up an external reference to a URN.  Preference is that it be registered with IANA, so we can have interoperability between systems.

**Comment by yonhan3** [comment]
I like gitoid:blob:sha1:${hash}

similarly, for bom_id we can have:

gitoid:bom:sha1:${hash}

**Comment by AevaOnline** [comment]
I can see a lot of value in 'gitoid:{header}:{hash function}' as a standard.

For reference / since I asked the question on our call a few minutes ago: the IANA doc for 'git' explains why that is not a suitable identifier here, namely that it is a URL to a git repository and not a reference to a git object id: https://www.iana.org/assignments/uri-schemes/prov/git 

I suggest we coordinate with the git community to see if they would also value this as a standard. If / when we want to move forward with IANA, I am happy to follow up with Dave Thaler.

**Comment by edwarnicke** [comment]
> I like gitoid:blob:sha1:${hash}
> 
> similarly, for bom_id we can have:
> 
> gitoid:bom:sha1:${hash}

@yonhan3 I'd been thinking of this as `gitoid:${git object type}:${hash type}:${hashvalue}` like `gitoid:blob:sha1:261eeb9e9f8b2b4b0d119366dda99c6fd7d35c64`.  

Where git object types are:
- blob - representing the contents of any file
- tree - a representation of a directory
- commit
- tag

So far we've been talking about treating the GitBOM docs as 'blobs' for purposes of computing the id meaning:

`sha1(blob ${size}\0${GitBOM document contents})`

were you suggesting we introduce a new git object type?

**Comment by yonhan3** [comment]
@edwarnicke No. I am not suggesting to introduce a new git object type. The gitBOM bom_id is the same calculation type of blob actually, so no need for gitoid:bom:sha1:${hash} at all. That is my mis-conception of bom_id as a different type than blob_id. Now it is clarified. Thanks.

**Comment by edwarnicke** [comment]
@yonhan3 All good!  Just wanted to make sure I understood you correctly.  The world is full of interesting ideas, I prefer to hear them than to make presumptions about them :)

**Comment by zvr-intel** [comment]
I strongly suggest to read on [SoftWare Heritage persistent IDentifiers (SWHIDs)](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html) before trying to re-invent something similar ;-).

They are a IANA-registered URI scheme, and they have started the process of becoming an ISO standard. If something is not already covered, I am sure extensions (for example, new object types) will be welcome.

Let me know if I can help in any way.

**Comment by edwarnicke** [comment]
@zvr-intel See above, we have read them, and there's some discussion above about them :)

**Comment by jeff-schutt** [comment]
^ issue suggests adding whatever GitBOM URI Namespace Identifier is decided on here as a PURL type, that way GitBOM Object Identifiers come "for free" when implementing PURL with an SBOM spec like CycloneDX or SPDX.

**Comment by jeff-schutt** [comment]
I like the flexibility that specifying the syntax in the standard as: `gitoid:${git object type}:${hash type}:${hashvalue}` provides for others to use gitbom identifiers in ways that haven't yet been considered.

**Comment by edwarnicke** [comment]
@jeff-schutt @kestewart Looking at [PURL](https://github.com/package-url/purl-spec) it seems oriented to identifying 'packages' not necessarily 'artifacts'.  Am I reading that right?

**Comment by edwarnicke** [comment]
After chatting with some of my friends in standards, they suggested filling out the [IANA template](https://www.rfc-editor.org/rfc/rfc7595.html#section-7) for gitoid and discussing it.

Here's a first pass at filling in the template for discussion.

```
Scheme name: gitoid

Status: provisional

Applications/protocols that use this scheme name:
     See [Section 3.5](https://www.rfc-editor.org/rfc/rfc7595.html#section-3.5).
Contact:
     Person (including contact information) to contact for further
     information.

Change controller:
     Organization or person (often the author), including contact
     information, authorized to change this.

References:
https://git-scm.com/book/en/v2/Git-Internals-Git-Objects

Scheme syntax: gitoid":"<git object type>":"<hash algorithm>":"<hash value>

Current valid values for <git object type> are 'blob','tree','commit', and 'tag'.
Current valid values for <hash algorithm> are 'sha1' and 'sha256'
<hash value> should be expressed as a hexidecimal string in lower case.

Example: gitoid:blob:sha1:261eeb9e9f8b2b4b0d119366dda99c6fd7d35c64

Scheme semantics:

gitoid stands for Git Object ID.  Git is an object store.  It currently has four types of objects.  A git object is formed by prepending an object header to the contents of the object.  A git object header consists of:

<git object type>" "<size of contents as decimal string>"\0"

The git object id is computed by applying a hash to the git object.  Currently the most common hash is sha1, but there is some effort to transition to sha256 over time.

A gitoid URI identifies a git object.  Given a byte array and a gitoid, it should be possible to construct the analogous git object header from the gitoid, as gitoid contains the git object type.  By prepending the constructed git object header to the byte array and performing the hash indicated in the gitoid, it should be possible to determine whether or not the gitoid matches the byte array.

Encoding considerations:
gitoid's should be encoded in utf-8, which for the range of valid characters is also ascii.

Interoperability considerations:
Unknown, use with care.

Security considerations:
Unknown, use with care.
```

**Comment by zacchiro** [comment]
This proposal seems to be entirely isomorphic to [SoftWare Heritage persistent IDentifiers (SWHIDs)](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html). See [examples](https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html#examples) from the same page.

SWHIDs are already:
- de facto standardized and used at large in the [Software Heritage archive](https://archive.softwareheritage.org/) to index the equivalent of ~20 billion git objects
- IANA registered (provisionally) under the ["swh:" URI prefix](https://www.iana.org/assignments/uri-schemes/prov/swh)
- supported in SPDX as alternative identifiers [since SPDX 2.2](https://www.linuxfoundation.org/blog/spdx-2-2-specification-released/)
- used as a [wikidata property](https://www.wikidata.org/wiki/Property:P6138)
- on their way to become an ISO standard (following the same path of SPDX)
- they are also already resolvable today using https://archive.softwareheritage.org/<SWHID> which would be a massive help to lookup objects in the context of SBOMs

Do we really need another isomorphic identifier scheme, just under a different name/prefix?

**Comment by edwarnicke** [comment]
@zacchiro I was perhaps less than clear above.  Let me try again :)

GitBOM needs flexibility over time wrt the computation of the gitoid.  The existing definition as I am reading it of swh:cnt is explicitly the sha1 gitoid.  Git has since moved to supporting a sha256 gitoid (though its slow so far to be adopted).

Additionally, effectively we are looking to specify a specific hash.  gitoid:blob:sha256 *does* specify an explicit hash.  It can evolve over time, it is bound semantically only to the fact its a git object id, of type blob, using sha256.  swh:cnt doesn't carry any of those semantics.

I am delighted however to hear you are in the process of designing SWHIDsv2 :)  

Would you consider shifting to something like:

swh:cnt:gitioid:blob:${hash type} which would bring you the semantic meaning (this is the hash used) and the flexibility overtime as git evolves that we need in GitBOM, while I suspect improving those characteristics in swh?

If you are interested in some of the discussion of the considerations on this matter, we have a discussion of this in the last GitBOM Spec meeting ( [link to exact start of discussion in our meeting recording on Youtube](https://www.youtube.com/watch?v=KwEXQZlvrzY) ).

**Comment by zvr** [comment]
To expand on @zacchiro comment: since both SWHID and GITOID are, by definition, compatible with git hashes, the actual hashes will be the same. The four different cases for GITOIDs ('blob','tree','commit', and 'tag') are already covered by SWHIDs.

One can have a simple 4-line script:
```
#!/usr/bin/env -S sed -f
# swhid2gitoid -- generate gitoid references
s/swh:1:cnt/gitoid:blob:sha1/g
s/swh:1:dir/gitoid:tree:sha1/g
s/swh:1:rev/gitoid:commit:sha1/g
s/swh:1:rel/gitoid:tag:sha1/g
```
and use it to process the examples from the SWHID documentation to produce correct results like:
> * `gitoid:blob:sha1:94a9ed024d3859793618152ea559a168bbcbb5e2` points to the content of a file containing the full text of the GPL3 license
> * `gitoid:tree:sha1:d198bc9d7a6bcf6db04f476d29314f157507d505` points to a directory containing the source code of the Darktable photography application as it was at some point on 4 May 2017
> * `gitoid:commit:sha1:309cf2674ee7a0749978cf8265ab91a60aea0f7d` points to a commit in the development history of Darktable, dated 16 January 2017, that added undo/redo supports for masks
> * `gitoid:tag:sha1:22ece559cc7cc2364edc5e5593d63ae8bd229f9f` points to Darktable release 2.3.0, dated 24 December 2016

**Comment by zvr** [comment]
Regarding sha256, I think saying "Git has since moved to supporting a sha256 gitoid" is a little... premature.
Yes, the code is there, but even the current version of git still comes with a big fat [warning](https://github.com/git/git/blob/master/Documentation/object-format-disclaimer.txt) in the documentation :

> THIS OPTION IS EXPERIMENTAL! SHA-256 support is experimental and still
> in an early stage.  A SHA-256 repository will in general not be able to
> share work with "regular" SHA-1 repositories.  It should be assumed
> that, e.g., Git internal file formats in relation to SHA-256
> repositories may change in backwards-incompatible ways.  Only use
> `--object-format=sha256` for testing purposes.

Not to mention that the platform we're currently using for communicating does not yet support such hashes ;-)

Nevertheless, I believe we all understand and agree on the importance of being future-proof; that's why SWHID has an explicit _version_ component in its format. Since the goal is to assist software archiving, it's not only sha1 and sha256 hashes that have to be supported, but whatever will be in use in the next years, decades, and centuries.

The current definition of SWHID says that _version 1_ only uses sha1, since this it the hash algorithm used in git. There is nothing stopping us defining next versions differently.

BTW, if it were only about differentiating between sha1/sha256, doesn't the length of the hash clearly show which algorithm has been used?

**Comment by zacchiro** [comment]
Hi @edwarnicke, thanks for the extra context/comments, which were very useful for me.

You are correct in mentioning that SWHIDv1 uses sha1. I wouldn't say that it uses sha1 *implicitly* though, in the sense that that hash system is bound to the version (currently 1) which is explicitly exposed on SWHIDs.

For SWHIDv2 we will definitely switch to sha256, although it is not clear yet if we will want to have git compatibility (I can elaborate on this, but in short the git manifest format has *a lot* of quirks, and having operated the largest "git archive" out there we have encountered our share of them…). I'd love to hear if that is a strong requirement for you or not. I notice that you are currently not compatible with git either, but I'm assuming you are counting on the fact that you *will* be git compatible when git also switched to sha256. Is that correct?

The other difference that seems to make what you plan to have incompatible with SWHIDs is that you desire having the hash name explicitly stated in the identifiers. Can you elaborate on why do you think that is useful? In my opinion, it is has useful has having an explicit specification version number (whit the hash defined in the specification); no more, no less. One might think that the hash name in the IDs is enough to allow in the future some form of "auto discovery" of the right hash algorithm, but that stops being true as soon as you look under the cover.
For instance, while we are talking about sha1 here, they are not at all sha1 computed on raw file contents; but rather "git style sha1" computed on a string formed with the git header + the raw file content. The same applies to sha256 as you are using them, as I understand it. So it does not seem realistic to imagine autodiscovery will work in the future; you will always need an external specification that describes how to interpret hash names and, at that point, a version number should be as good. Again, maybe I'm missing something of your design goals here. In that case I apologize and I'd love to be pointed to them.

Zooming out a bit, in my opinion the main advantage of using SWHIDs for gitbom would be archival + resolvability. Imagine receiving a random SBOM from someone and being able to click on any ID on it to lookup the file content and/or download the directory corresponding to it, without having to receive out-of-band the complete/corresponding source code from the same entity that has given you the SBOM.
This is a functionality that we can currently offer via Software Heritage (limited to what we have archived of course, but we also do on-demand archival, so that is not a major blocker and we can imagine further integration with gitbom workflows).
But we can not offer that for all identifier schemes that exist out there, because each one will come with keeping an additional association table for tens of billion files.

Again, I do not if this archival+resolvability feature is a goal for you, but I'd love to understand how we can make it work for gitbom if it is of interest for you.

Thanks a lot for engaging in this discussion, I really appreciate what you are trying to do, and clearly our goals have a lot in common.

**Comment by dpp** [comment]
It strikes me that both GitBOM and SWHID are addressing significantly overlapping problem spaces using very similar approaches: DAGs of "stuff" that make up a system in a Merkle-like manner (DAG, tree, ...)

Here are some of the differences I see (note that today is the first day I've looked at the SWHID stuff, so I will have missed a bunch of stuff and hope for corrections/education):

* The SWHID is limited to a specific set of types: "snp" "rel" "rev" "dir"  "cnt". I am expecting that gitbom will have a richer and perhaps open set of types. In my mind:
    * there will be hashes of code snippets and those hashes will contained in files. As copy/pasting becomes automated (e.g., GitHub CoPilot), being able to trace snippets will become more important.
    * Things like compiler versions and compiler flags and other things that are part of converting source to executable (which may be multi-phased... e.g. Java 8-102 compiling a class file [a separate gitbom] that's packaged into a WAR file [a separate gitbom] which is executed by Java 17-672 [a separate gitbom] running on Ubuntu 22.04 on a 32 bit ARM chip etc.) (note... this is pure dpp-crazy talk... I do not represent Aeva or Ed's opinion here)
    * So... I see gitboms as turtles... and the SWHID spec does address the turtle/recursive nature of describing the components of a system
* The flexibility of specifying the hash algorithm as part of the URx (URI or URN) will pay benefits. A single artifact may be described with different hashes based on the different hashing algorithm. Now... if we use some kind of distributed hash tably kinda thing for lookup, maybe the hash algorithm becomes less important... we just look up the hash... but... dunno... it feels right to be explicit

I guess after writing this, it seems that the SWHID is a subset of what we will describe with GitBOM. As the SWHID team is moving to a v2 of the URx scheme, would it be possible to allow for the flexibility that the GitBOM project may need?

**Comment by alilleybrinker** [comment]
Closing, as I believe we did resolve this with the following:

- Addition of a URI spec to the main `spec` repository: https://github.com/omnibor/spec/blob/main/spec/GITOID_URI_SPEC.txt
- Registering the scheme with IANA (current status 'provisional'): https://www.[iana.org/assignments/uri-schemes/prov/gitoid](https://www.iana.org/assignments/uri-schemes/prov/gitoid)

If anyone feels there are further points to raise, it may better to open a fresh issue and reference any relevant conversations in this one.
